diffrax/_solver/kl.py

```diff
@@ -0,0 +1,291 @@
+import operator
+from typing import Optional
+
+import equinox as eqx
+import jax.tree_util as jtu
+import lineax as lx
+from jax import numpy as jnp
+from jaxtyping import Array, PyTree
+
+from .._custom_types import (
+    Args,
+    BoolScalarLike,
+    Control,
+    DenseInfo,
+    RealScalarLike,
+    VF,
+    Y,
+)
+from .._heuristics import is_sde
+from .._solution import RESULTS
+from .._term import (
+    AbstractTerm,
+    ControlTerm,
+    MultiTerm,
+    ODETerm,
+)
+from .base import (
+    _SolverState,
+    AbstractSolver,
+    AbstractWrappedSolver,
+)
+
+
+class KLState(eqx.Module):
+    """
+    The state of the SDE and the KL divergence.
+    """
+
+    y: Y
+    kl_metric: Array
+
+
+def _compute_kl_integral(
+    drift_term1: ODETerm,
+    drift_term2: ODETerm,
+    diffusion_term: ControlTerm,
+    t0: RealScalarLike,
+    y0: Y,
+    args: Args,
+    linear_solver: lx.AbstractLinearSolver,
+) -> KLState:
+    """
+    Compute the KL divergence.
+    """
+    drift1 = drift_term1.vf(t0, y0, args)
+    drift2 = drift_term2.vf(t0, y0, args)
+    drift = jtu.tree_map(operator.sub, drift1, drift2)
+
+    diffusion = diffusion_term.vf(t0, y0, args)  # assumes same diffusion
+
+    divergences = lx.linear_solve(diffusion, drift, solver=linear_solver).value
+
+    kl_divergence = jtu.tree_map(lambda x: 0.5 * jnp.sum(x**2), divergences)
+    kl_divergence = jtu.tree_reduce(operator.add, kl_divergence)
+
+    return KLState(drift1, jnp.squeeze(kl_divergence))
+
+
+class _KLDrift(AbstractTerm):
+    drift1: ODETerm
+    drift2: ODETerm
+    diffusion: ControlTerm
+    linear_solver: lx.AbstractLinearSolver
+
+    def vf(self, t: RealScalarLike, y: Y, args: Args) -> KLState:
+        y = y.y
+        return _compute_kl_integral(
+            self.drift1, self.drift2, self.diffusion, t, y, args, self.linear_solver
+        )
+
+    def contr(self, t0: RealScalarLike, t1: RealScalarLike, **kwargs) -> Control:
+        return t1 - t0
+
+    def prod(self, vf: VF, control: RealScalarLike) -> Y:
+        return jtu.tree_map(lambda v: control * v, vf)
+
+
+class _KLControlTerm(AbstractTerm):
+    control_term: ControlTerm
+
+    def vf(self, t: RealScalarLike, y: Y, args: Args) -> KLState:
+        y = y.y
+        vf = self.control_term.vf(t, y, args)
+        return KLState(vf, jnp.array(0.0))
+
+    def contr(self, t0: RealScalarLike, t1: RealScalarLike, **kwargs) -> KLState:
+        return KLState(self.control_term.contr(t0, t1), jnp.array(0.0))
```

Comment:
I don't think this needs to output a `KLState`. The output of this is only used as an input to `prod`/`vf_prod`, which we control.

Reply:
Understandable, changed